### ISSUE TITLE
Unable to inherit from MobileElement class for version 3.1.0

### ISSUE THREAD

**Issue by alewang** [appium/java-client#220]
I am unable to inherit from the MobileElement class. One of the samples of the errors I see is:

    - The return types are incompatible for the inherited methods
    FindsByLinkText.findElementsByLinkText(String),
    MobileElement.findElementsByLinkText(String)

I took a look through the source code, and it seems the cause is due to the DefaultGenericMobileElement class (MobileElement's parent class) extending from RemoteWebElements and implementing the various interfaces. The method signatures in RemoteWebElements and the Generic* interfaces match exactly, but they have different return types. RemoteWebElement's methods' return types are of the type List<WebElement>, while the return types for the various interfaces are List<T> (with T being the parametrized type).

I would like to extend from the MobileElement class for some very specific functionality. I tried using the @SuppressWarnings annotation with various names (including "all"), but nothing seems to tell Eclipse that this error is not a problem.

What can I do on my end to "fix" this error?

What can I do for Appium to prevent these conflicting, overlapping methods with equivalent signatures from well.... overlapping?

**Comment by TikhomirovSergey** [comment]
Hi @alewang
Right now I could suggest you to use Intellij Idea 12.x or higher. You have to enable Eclipse java compiler. Also I could suggest you to add the folowing at your pom file:

```xml
	<build>
		<plugins>
			...
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.3</version>
				<configuration>
					<source>1.7</source>
					<target>1.7</target>
                    <compilerId>eclipse</compilerId>
				</configuration>
                <dependencies>
                    <dependency>
                        <groupId>org.codehaus.plexus</groupId>
                        <artifactId>plexus-compiler-eclipse</artifactId>
                        <version>2.5</version>
                    </dependency>
                </dependencies>
			</plugin>
                 ...
		</plugins>
	</build>
```

If it is not acceptable solution and it extremally needs to use eclipse then it needs time. Which Eclipse version do you use? I use Eclipse Luna. There are many plugins are installed for my needs (Spring, JBoss tools and more). When I use this Eclipse instance the code like yours is compiled. When I use another instance without these plugins I'm facing the same problem. So one of plugins or their companents manages this code to be compiled. So there is lack of information abou Eclipse configuration at README (this info is for contributors).


Why do you have to extend MobileElement? There are two interesting points. The first is that it is possible to find more simple solution. The second is.. I think about a new feature for Appium java client. Maybe it is your case. :)

**Comment by TikhomirovSergey** [comment]
I'm working on your problem...

**Comment by alewang** [comment]
Hi @TikhomirovSergey!

Thanks for the help. I will give this a try today.

**Comment by alewang** [comment]
Our reason for why we want to extend MobileElement is that we want to cache some data about an element so that we do not have to query the Appium server for that data each time. In one of our particular cases, it was most natural to have the element object itself maintain a reference to its own cached data, so we tried doing that by extending the MobileElement class. We saw an interesting surprise when we changed the java-client version to 3.1.0.

**Comment by TikhomirovSergey** [comment]
Yeah! It is compiled some unusual way in order to avoid excessive casting from WebElement to MobileElement. Usesrs complained so much. But java has some restrictions. 

Why do you not use page objects? [Page object design pattern](https://code.google.com/p/selenium/wiki/PageObjects). [Page factory for java client](https://code.google.com/p/selenium/wiki/PageFactory).

**Comment by alewang** [comment]
I'll take a look! Thanks for the help.

**Comment by TikhomirovSergey** [comment]
Do IDEA and it's settings help you?

It is not possible to solve this problem only on Appium side. Fortunately I have a good idea... ;)

**Comment by afwang** [comment]
I tried IDEA 14.1.4 with the updated settings in the POM file, but the same error shows up on the child class.

I'm thinking about ways to solve this, and I think the cleanest solution might be for me to go to Selenium's source and open a pull request to make their RemoteWebElements class parametrized. When I have some time, I'll try doing that.

**Comment by TikhomirovSergey** [comment]
Have you tried to change java compiler to "Eclipse" in IDEA?

>Right now I could suggest you to use Intellij Idea 12.x or higher. You have to enable Eclipse java compiler. Also I could suggest you to add the folowing at your pom file:

**Comment by afwang** [comment]
Neat, I didn't know I could change the compiler. The error just shows as a warning with the Eclipse compiler.

**Comment by TikhomirovSergey** [comment]
For somebody who doesn't know how to change compiler
1. Go to "Settings"
2. 
![compiler](https://cloud.githubusercontent.com/assets/4927589/8992783/266c2346-3709-11e5-8d14-82de4324c666.png)

@afwang It will just allow to compile a subclass.

**Comment by TikhomirovSergey** [comment]
Please look at https://github.com/SeleniumHQ/selenium/pull/863

**Comment by TikhomirovSergey** [comment]
Fixed at #413
Will be published at 4.1.0
